Add Thinkpad NXP1001 NFC support using libnfc-nci and PC/SC

StarGate01 opened the pull request:
This PR adds a new NFC stack based on the NXP libnfc-nci userland, an ACPI-configured kernel driver, and a pcscd-lite module for easy use.

This module specifically targets the NFC chipset commonly found in Lenovo Thinkpads, the NXP `NXP1001` module. Unfortunately this chipset has no proper support in either `libnfc` or `neard` (i.e. the upstream Linux kernel).

The implementation and configuration has been engineered in a community effort over a few years. For details, please refer to https://github.com/nfc-tools/libnfc/issues/455 , specifically https://github.com/nfc-tools/libnfc/issues/455#issuecomment-2210704361 , https://github.com/nfc-tools/libnfc/issues/455#issuecomment-2221979571 , and https://github.com/nfc-tools/libnfc/issues/455#issuecomment-2229756225 for my implementation.

The PR adds the following new packages / modules:
- **linuxPackages.nxp-pn5xx: init at 0.4-unstable**: The kernel driver, sourced from https://github.com/jr64/nxp-pn5xx , which in turn is a fork of the original sources by NXP (https://github.com/NXPNFCLinux/nxp-pn5xx), with the added capability of automatic hardware configuration via the ACPI tables, and updates for compilation with modern recent kernel version.
- **libnfc-nci: init at 2.4.1-unstable**: The userland library which interfaces the kernel driver, sourced from https://github.com/StarGate01/linux_libnfc-nci/ . This is a fork of the original NXP library (https://github.com/NXPNFCLinux/linux_libnfc-nci) with quite a few patches and configurations from the community and myself, in order to make the library work properly for the Lenovo chipset.
- **ifdnfc-nci: init at 0.2.1**: The pcsc-lite IFD handler to expose the NFC chip as a PC/SC compliant reader for APDU exchange. Sourced from https://github.com/StarGate01/ifdnfc-nci , uses the modified libnfc-nci library. This enables other applications like Firefox, Yubico Authenticator, and KeePassXC to use the NFC chip as NFC a smartcard reader.
- **pcscd: allow multiple readerConfig entries**: Does what it says. This is used by the nixos module to automatically add the IFD handler to the pscs-lite configuration file.
- **nixos/nfc-nci: init**: Adds a new NixOS module to enable and configure this entire stack et once via `config.hardware.nfc-nci`. It sets up:
  - Blacklisted (conflicting) kernel modules: `nxp_nci_i2c` and `nxp_nci` (default, configurable via `config.hardware.nfc-nci.blacklistedKernelModules`
  - libnfc-nci required configuration files at `/etc/libnfc-{nci,nxp-init,nxp-pn547,nxp-pn548}.conf`, with the values taken from `config.hardware.nfc-nci.settings.{nci,init,pn54x}`, respectively. The default values provide a sane configuration for the Lenovo module.
  - Udev rules for the `/dev/pn544` device node, this rule is exposed by the kernel package.
  - Add the kernel module to the list of loaded modules and enable it
  - Set up the pcsc-lite IFD handler, this can be controlled by `config.hardware.nfc-nci.enableIFD`.
  - Set up a systemd rule to recover the NFC module from hibernation-induced powerdown

I do realize this is quite a big PR, and I invite the community and also all the people from https://github.com/nfc-tools/libnfc/issues/455 to critique and improve this PR. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

StarGate01 committed:
libnfc-nci: init at 2.4.1-unstable

StarGate01 committed:
linuxPackages.nxp-pn5xx: init at 0.4-unstable

StarGate01 committed:
nixos/nfc-nci: init

StarGate01 committed:
pcscd: allow multiple readerConfig entries

StarGate01 committed:
ifdnfc-nci: init at 0.2.1

SuperSandro2000 commented on pkgs/by-name/if/ifdnfc-nci/package.nix:
```diff
@@ -0,0 +1,39 @@
+  src = fetchFromGitHub {
+    owner = "StarGate01";
+    repo = "ifdnfc-nci";
+    rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

SuperSandro2000 commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+      '';
+      NXP_NFC_PROPRIETARY_CFG = "{05:FF:FF:06:81:80:70:FF:FF}";
+      NXP_EXT_TVDD_CFG = "0x01";
+      NXP_EXT_TVDD_CFG_1 = ''
+        {20, 02, 07, 01,
+                A0, 0E, 03, 16, 09, 00
+              }'';
```
```suggestion
      NXP_EXT_TVDD_CFG_1 = ''
        {20, 02, 07, 01,
                A0, 0E, 03, 16, 09, 00
              }
      '';
```

SuperSandro2000 commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+      NXP_NFC_PROFILE_EXTN = ''
+        {20, 02, 05, 01,
+            A0, 44, 01, 00
+              }'';
```
```suggestion
              }
      '';
```

SuperSandro2000 commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+    blacklistedKernelModules = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [
+        "nxp_nci_i2c"
```
We should preserve this when setting the option, shouldn't we? Also does this need to be an option at all?

StarGate01 replied:
I agree. This is probably a leftover from testing, I have removed the config option, the module now always adds these two kernel modules to the blacklist.

SuperSandro2000 commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+      "nxp-pn5xx"
+    ];
+
+    services.pcscd.readerConfigs =
+      [ ]
+      ++ lib.optionals cfg.enableIFD [
```
```suggestion
    services.pcscd.readerConfigs = lib.mkIf cfg.enableIFD [
```

SuperSandro2000 commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+
+    # NFC chip looses power when system goes to sleep / hibernate,
+    # and needs to be re-initialized upon wakeup
+    powerManagement.resumeCommands =
+      ''''
+      + lib.optionals cfg.enableIFD ''
```
```suggestion
    powerManagement.resumeCommands = lib.optionalString cfg.enableIFD ''
```

misuzu commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+    };
+
+    services.udev.packages = [
+      pkgs.linuxPackages.nxp-pn5xx
```
```suggestion
      config.boot.kernelPackages.nxp-pn5xx
```

misuzu commented on nixos/modules/hardware/nfc-nci.nix:
```diff
@@ -0,0 +1,168 @@
+
+    boot.blacklistedKernelModules = cfg.blacklistedKernelModules;
+    boot.extraModulePackages = [
+      pkgs.linuxPackages.nxp-pn5xx
```
```suggestion
      config.boot.kernelPackages.nxp-pn5xx
```

misuzu commented on nixos/modules/module-list.nix:
```diff
@@ -85,6 +85,7 @@
   ./hardware/network/eg25-manager.nix
   ./hardware/network/intel-2200bg.nix
   ./hardware/new-lg4ff.nix
+  ./hardware/nfc-nci.nix
```
Needs a release notes entry

StarGate01 commented:
Thank you @SuperSandro2000 and @misuzu  for your helpful suggestions, I have implemented them all. I have also added an entry in the release notes and cleaned up the syntax a bit. In addition I have added a lot of code comments to explain the NXP NCI configuration values in the module.

misuzu commented on pkgs/os-specific/linux/nxp-pn5xx/default.nix:
```diff
@@ -0,0 +1,41 @@
+    license = licenses.gpl2Only;
+    maintainers = with maintainers; [ stargate01 ];
+    platforms = platforms.linux;
+  };
```
Doesn't build on 6.13.
```suggestion
    broken = kernel.kernelAtLeast "6.13";
  };
```

StarGate01 replied:
Thank you for bringing this up, this is a upstream problem with the kernel module which I will address before merging this PR.

StarGate01 replied:
Thanks to the upstream module patch, the module now properly builds and runs on kernel 6.13 as well. The issue was the (legitimate) coexistence of `CONFIG_OF` and `CONFIG_ACPI` in the kernel config, the module now no longer uses compile time configuration but instead properly does runtime devicetree / ACPI detection.

misuzu commented on pkgs/os-specific/linux/nxp-pn5xx/default.nix:
```diff
@@ -0,0 +1,41 @@
+    echo 'SUBSYSTEM=="misc", KERNEL=="pn544", MODE="0666", GROUP="dialout"' > $out/etc/udev/rules.d/99-nxp-pn5xx.rules
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

misuzu commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 379118`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_5_4_hardened.nxp-pn5xx</li>
    <li>linuxPackages_5_4_hardened.nxp-pn5xx</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>linuxPackages_latest.nxp-pn5xx (linuxKernel.packages.linux_6_13.nxp-pn5xx)</li>
    <li>linuxPackages_latest-libre.nxp-pn5xx (linuxKernel.packages.linux_latest_libre.nxp-pn5xx)</li>
    <li>linuxPackages_zen.nxp-pn5xx (linuxKernel.packages.linux_zen.nxp-pn5xx)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 19 packages built:</summary>
  <ul>
    <li>ifdnfc-nci</li>
    <li>libnfc-nci</li>
    <li>linuxKernel.packages.linux_5_10.nxp-pn5xx</li>
    <li>linuxPackages_5_10_hardened.nxp-pn5xx (linuxKernel.packages.linux_5_10_hardened.nxp-pn5xx)</li>
    <li>linuxKernel.packages.linux_5_15.nxp-pn5xx</li>
    <li>linuxPackages_5_15_hardened.nxp-pn5xx (linuxKernel.packages.linux_5_15_hardened.nxp-pn5xx)</li>
    <li>linuxKernel.packages.linux_5_4.nxp-pn5xx</li>
    <li>linuxKernel.packages.linux_6_1.nxp-pn5xx</li>
    <li>linuxKernel.packages.linux_6_11.nxp-pn5xx</li>
    <li>linuxPackages_6_11_hardened.nxp-pn5xx (linuxKernel.packages.linux_6_11_hardened.nxp-pn5xx)</li>
    <li>linuxPackages.nxp-pn5xx (linuxKernel.packages.linux_6_12.nxp-pn5xx)</li>
    <li>linuxPackages_hardened.nxp-pn5xx (linuxPackages_6_12_hardened.nxp-pn5xx)</li>
    <li>linuxPackages_6_1_hardened.nxp-pn5xx (linuxKernel.packages.linux_6_1_hardened.nxp-pn5xx)</li>
    <li>linuxKernel.packages.linux_6_6.nxp-pn5xx</li>
    <li>linuxPackages_6_6_hardened.nxp-pn5xx (linuxKernel.packages.linux_6_6_hardened.nxp-pn5xx)</li>
    <li>linuxPackages-libre.nxp-pn5xx (linuxKernel.packages.linux_libre.nxp-pn5xx)</li>
    <li>linuxPackages_lqx.nxp-pn5xx (linuxKernel.packages.linux_lqx.nxp-pn5xx)</li>
    <li>linuxPackages_xanmod.nxp-pn5xx (linuxKernel.packages.linux_xanmod.nxp-pn5xx)</li>
    <li>linuxPackages_xanmod_latest.nxp-pn5xx (linuxKernel.packages.linux_xanmod_latest.nxp-pn5xx ,linuxPackages_xanmod_stable.nxp-pn5xx)</li>
  </ul>
</details>

misuzu commented on pkgs/os-specific/linux/nxp-pn5xx/default.nix:
```diff
@@ -0,0 +1,41 @@
+    description = "NXP's NFC Open Source Kernel mode driver with ACPI configuration support";
+    homepage = "https://github.com/jr64/nxp-pn5xx";
+    license = licenses.gpl2Only;
+    maintainers = with maintainers; [ stargate01 ];
```
```suggestion
    maintainers = with lib.maintainers; [ stargate01 ];
```

misuzu commented on pkgs/by-name/li/libnfc-nci/package.nix:
```diff
@@ -0,0 +1,47 @@
+    rm -rf $out/etc
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

misuzu commented on pkgs/by-name/li/libnfc-nci/package.nix:
```diff
@@ -0,0 +1,47 @@
+    description = "Linux NFC stack for NCI based NXP NFC Controllers";
+    homepage = "https://github.com/NXPNFCLinux/linux_libnfc-nci";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ stargate01 ];
```
```suggestion
    maintainers = with lib.maintainers; [ stargate01 ];
```

misuzu commented on pkgs/by-name/if/ifdnfc-nci/package.nix:
```diff
@@ -0,0 +1,39 @@
+    libnfc-nci
+  ];
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

misuzu commented on pkgs/by-name/if/ifdnfc-nci/package.nix:
```diff
@@ -0,0 +1,39 @@
+    homepage = "https://github.com/StarGate01/ifdnfc-nci";
+    license = licenses.gpl3Only;
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ stargate01 ];
```
```suggestion
    maintainers = with lib.maintainers; [ stargate01 ];
```
